File: programmers/level2/후보키.py
from itertools import combinations


def solution(relation):
    answer = []
    columnLength = len(relation[0])
    rowLength = len(relation)
    column = [i for i in range(columnLength)]

    keyList = []
    keyDict = dict()
    for i in range(columnLength):
        keyDict[i] = 0

    for i in range(1, columnLength):
        for comb in combinations(column, i):
            keySet = set()
            for row in relation:
                subKey = tuple(row[col] for col in comb)
                keySet.add(subKey)
            if len(keySet) == rowLength:
                keyList.append(comb)
    # 최소성 검사
    for k in keyList:
        isMinimal = True
        for existingKey in answer:
            if set(existingKey).issubset(
                set(k)
            ):  # 기존 후보키가 현재 키의 부분집합이면 제외
                isMinimal = False
                break
        if isMinimal:
            answer.append(k)

    if len(answer) == 0:
        return 1
    return len(answer)

File: programmers/level2/후보키.js
function solution(relation) {
  let answer = 0;
  const colNum = relation[0].length;
  const col = [];
  for (let i = 0; i < colNum; i++) {
    col.push(i);
  }
  let combinations = [];
  for (let i = 1; i <= colNum; i++) {
    const comb = combination(col, i);
    combinations.push(...comb);
  }
  combinations = checkUniqueness(relation, combinations);
  answer = checkDuplications(combinations).filter((e) => e).length;

  return answer;
}

function checkUniqueness(relation, combinations) {
  const results = []; // 유일성을 만족하는 조합

  combinations.forEach((combination) => {
    const set = new Set();
    relation.forEach((rel) => {
      set.add(combination.map((combi) => rel[combi]).join(","));
    });
    if (set.size == relation.length) {
      results.push(combination);
    }
  });
  return results;
}

function isSubset(setA, setB) {
  for (let elem of setA) {
    if (!setB.has(elem)) {
      return false;
    }
  }
  return true;
}

function checkDuplications(combinations) {
  const checkList = new Array(combinations.length).fill(true);
  const setCombi = combinations.map((e) => new Set(e));
  for (let i = 0; i < combinations.length; i++) {
    if (!checkList[i]) {
      continue;
    }
    for (let j = i + 1; j < combinations.length; j++) {
      if (isSubset(setCombi[i], setCombi[j])) {
        checkList[j] = false;
      }
    }
  }
  return checkList;
}

function combination(inputArray, num) {
  const result = [];

  function recur(arr, n) {
    if (arr.length === num) {
      result.push(arr);
      return;
    }
    for (let i = n; i < inputArray.length; i++) {
      recur([...arr, inputArray[i]], i + 1);
    }
  }
  recur([], 0);
  return result;
}
